src/device/wg_controller.rs

```diff
@@ -250,27 +246,52 @@ fn check_allowed_ips_correctness(peers: &BTreeMap<PublicKey, RequestedPeer>) ->
         .ok_or_else(|| Error::BadAllowedIps.into())
 }
 
-async fn consolidate_firewall<W: WireGuard, F: Firewall>(
+async fn consolidate_firewall<F: Firewall>(
     requested_state: &RequestedState,
-    wireguard_interface: &W,
     firewall: &F,
 ) -> Result {
-    let from_peers = wireguard_interface.get_interface().await?.peers;
-    let from_keys: HashSet<PublicKey> = from_peers.keys().copied().collect();
-    let to_peers: Vec<&telio_model::config::Peer> = requested_state
+    let from_keys_peer_whitelist: HashSet<PublicKey> =
+        firewall.get_peer_whitelist().iter().copied().collect();
+    let from_keys_ports_whitelist: HashSet<PublicKey> =
+        firewall.get_port_whitelist().keys().copied().collect();
+    let mut to_peers: HashMap<PublicKey, Node> = requested_state
         .meshnet_config
         .iter()
         .flat_map(|c| c.peers.iter())
         .flatten()
+        .map(|p| (p.public_key, p.into()))
         .collect();
-    let to_keys: HashSet<PublicKey> = to_peers.iter().map(|p| p.public_key).collect();
-    let delete_keys = &from_keys - &to_keys;
+    if let Some(exit_node) = &requested_state.exit_node {
+        if let Some(peer) = to_peers.get_mut(&exit_node.public_key) {
+            peer.is_exit = true;
+        } else {
+            to_peers.insert(exit_node.public_key, exit_node.try_into()?);
+        }
+    }
+    let to_keys: HashSet<PublicKey> = to_peers.keys().copied().collect();
+
+    let delete_keys = &from_keys_peer_whitelist - &to_keys;
     for key in &delete_keys {
         firewall.remove_from_peer_whitelist(*key);
+    }
+
+    let delete_keys = &from_keys_ports_whitelist - &to_keys;
+    for key in &delete_keys {
         firewall.remove_from_port_whitelist(*key);
     }
-    for node in to_peers {
-        firewall_upsert_node(firewall, &node.into());
+
+    for node in to_peers.values() {
+        if node.allow_incoming_connections || node.is_exit {
+            firewall.add_to_peer_whitelist(node.public_key);
+        } else {
+            firewall.remove_from_peer_whitelist(node.public_key);
+        }
+
+        if node.allow_peer_send_files {
+            firewall.add_to_port_whitelist(node.public_key, FILE_SEND_PORT);
+        } else {
+            firewall.remove_from_port_whitelist(node.public_key);
+        }
```

Comment:
@jjanowsk What are you thoughts about this:

Instead of building `to_peers` which is a list of Nodes, we build two sets of peers which need to be peer-whitelisted or port-whitelisted directly. 

It could look somewhat similar to this:
```rust
    let from_keys_peer_whitelist: HashSet<PublicKey> =
        firewall.get_peer_whitelist().iter().copied().collect();
    let from_keys_ports_whitelist: HashSet<PublicKey> =
        firewall.get_port_whitelist().keys().copied().collect();

    // Build a list of peers expected to be peer-whitelisted according
    // to allow_incoming_connections permission
    let to_keys_peer_whitelist: HashSet<PublicKey> = requested_state
        .meshnet_config
        .iter()
        .flat_map(|c| c.peers.iter())
        .flatten()
        .filter(|p| p.allow_incoming_connections)
        .map(|p| p.public_key)
        .collect();

    // VPN peer must always be peer-whitelisted
    if let Some(exit_node) = &requested_state.exit_node {
        //TODO
    }


    // Build a list of peers expected to be port-whitelisted according
    // to allow_peer_send_files permission
    let to_keys_ports_whitelist: HashSet<PublicKey> = requested_state
        .meshnet_config
        .iter()
        .flat_map(|c| c.peers.iter())
        .flatten()
        .filter(|p| p.allow_peer_send_files)
        .map(|p| p.public_key)
        .collect();

    // Consolidate peer-whitelist
    let delete_keys = &from_keys_peer_whitelist - &to_keys_peer_whitelist;
    for key in delete_keys {
        firewall.remove_from_peer_whitelist(key);
    }
    let add_keys = &to_keys_peer_whitelist - &from_keys_peer_whitelist;
    for key in add_keys {
        firewall.add_to_peer_whitelist(key);
    }

    // Consolidate port-whitelist
    let delete_keys = &from_keys_ports_whitelist - &to_keys_ports_whitelist;
    for key in delete_keys {
        firewall.remove_from_port_whitelist(key);
    }
    let add_keys = &to_keys_ports_whitelist - &from_keys_ports_whitelist;
    for key in add_keys {
        firewall.add_to_port_whitelist(key, FILE_SEND_PORT);
    }
```

Reply:
I agree. It looks better. I'll apply this